lib/hashie/extensions/mash/maintain_original_keys.rb

```diff
@@ -0,0 +1,52 @@
+module Hashie
+  module Extensions
+    module Mash
+      # Overrides the indifferent access of a Mash to maintain keys in the
+      # original format given to the Mash.
+      #
+      # @example
+      #   class MaintainingMash < Hashie::Mash
+      #     include Hashie::Extensions::Mash::MaintainOriginalKeys
+      #   end
+      #
+      #   mash = MaintainingMash.new(:symbol_key => :symbol, 'string_key' => 'string')
+      #   mash.to_hash  #=> { :symbol_key => :symbol, 'string_key' => 'string' }
+      #   mash['string_key'] == mash[:string_key]  #=> true
+      #   mash[:symbol_key] == mash['symbol_key']  #=> true
+      module MaintainOriginalKeys
+        private
+
+        def self.included(descendant)
+          unless descendant <= Hashie::Mash
+            fail ArgumentError, "#{descendant} is not a kind of Hashie::Mash"
+          end
+        end
+
+        # Converts symbol/string keys to their alternative formats, but leaves
+        # other keys alone.
+        #
+        # @param [Object, String, Symbol] key the key to convert.
+        # @return [Object, String, Symbol] the converted key.
+        def convert_key(key)
+          case key
+          when Symbol then key.to_s
+          when String then key.to_sym
+          else key
+          end
+        end
+
+        # Indifferently accesses a key in the Mash.
+        #
+        # @param [Object, String, Symbol] key the key to access.
+        # @return [Object] the value assigned to the key.
+        def indifferent_key(key)
```

Comment:
I believe that all you need is an overriden `convert_key` calling `super`. I don't see any benefit in introducing another notation, it only complicates things.